lib/elixir/lib/code.ex

```diff
@@ -1197,6 +1197,188 @@ defmodule Code do
     :elixir.string_to_quoted!(to_charlist(string), line, column, file, opts)
   end
 
+  @doc """
+  Converts the given string to its quoted form and a list of comments.
+
+  This function is useful when performing textual changes to the source code,
+  while preserving information like comments and literals position.
+
+  Returns `{:ok, quoted_form, comments}` if it succeeds,
+  `{:error, {line, error, token}}` otherwise.
+
+  The AST returned differs from regular AST in that literals are wrapped in
+  a block to preserve their metadata. For example, the literal `"foo"` will be
+  converted to `{:__block__, [line: 1, delimiter: "\"", token: "\"foo\""], ["foo"]}`.
+
+  In some cases the wrapping is not uniform, for example in the case of keyword
+  lists the wrapping may change if the square were ommitted:
+      iex> Code.string_to_quoted_with_comments("[foo: :bar]")
+      {:ok, {
+        :__block__, [closing: [line: 1], line: 1], [[
+          {{:__block__, [format: :keyword, line: 1], [:foo]}, {:__block__, [line: 1], [:bar]}}
+        ]]
+      }, []}
+
+      iex> Code.string_to_quoted_with_comments("a when foo: :bar")
+      {:ok, {
+        :when, [line: 1], [
+          {:a, [line: 1], nil},
+          [
+            {{:__block__, [format: :keyword, line: 1], [:foo]}, {:__block__, [line: 1], [:bar]}}
+          ]
+        ]
+      }, []}
+
+  Note that in the second example the outer wrapping of the keyword list is
+  omitted.
+
+  Comments are maps containing information about the line they were found, its
+  contents, and how many end of lines were found between the comment and the
+  closest tokens:
+      iex> Code.string_to_quoted_with_comments("\""
+      ...> :foo
+      ...>
+      ...> # Hello, world!
+      ...>
+      ...>
+      ...> # Some more comments!
+      ...> "\"")
+      {:ok, {:__block__, [line: 1], [:foo]}, [
+        %{line: 3, previous_eol: 2, next_eol: 3, text: "\# Hello, world!"},
+        %{line: 6, previous_eol: 3, next_eol: 1, text: "\# Some more comments!"},
```

Comment:
In some cases previous_eol can be nil, right? Do you know when? Should we make it default to zero instead? 🤔

Reply:
I think it's already defaulting to `2` in line 1354?

Reply:
I see. I think gather_comments and format_comment should remain in the formatter because they are formatter specific behaviour. The code itself should return the comments as close to how it was parsed as possible. WDYT?

And then we make previous_eol default to 0 and convert 0 (instead of nil) to 2.

That's the tricky part about these changes, we need to export all of those with values that make sense outside of the formatter, so we can't take as many shortcuts. :)

Reply:
I agree, I will get to it now :)

Reply:
Done!

> Do you know when?
It seems to happen when the comment is in the same line as the code, for example:
```elixir
foo # Hello!
```

Reply:
Ah, so it should be 0 indeed. Perfect.